### PR TITLE
Start to incorporate data from the maven lock into our SBOM

### PR DESCRIPTION
**Pull request by aiuto**
This works so far, but oh what hackery. Bazel uses a non-standard way of referencing maven artifacts, so the external references we see don't map back to the lock file directly. And the lock file is a tricky beast to understand.

I am happy I have a few elements in place.  I will come back for more in the next pass.

### CODE REVIEW

File: tools/compliance/gather_packages.bzl [hunk 1]
```diff
@@ -154,9 +153,15 @@ def gather_package_common(target, ctx, provider_factory, metadata_providers, fil
...
     # Record all the external repos anyway.
     target_name = str(target.label)
+    packages = None
     if target_name.startswith("@") and target_name[1] != "/":
         packages = [target.label]
         # DBG print(str(target.label))
+    elif hasattr(ctx.rule.attr, "tags"):
+        for tag in ctx.rule.attr.tags:
+            if tag.startswith("maven_coordinates="):
+                print(tag)
```
Comment (sdtwigg) [inline tools/compliance/gather_packages.bzl:163]: probably get rid of this line

File: tools/compliance/sbom.bzl [hunk 2]
```diff
@@ -41,7 +47,8 @@ _sbom = rule(
     implementation = _sbom_impl,
     attrs = {
         "packages_used": attr.label(
-            allow_files = True,
+            #allow_files = True,
```
Comment (sdtwigg) [inline tools/compliance/sbom.bzl:50]: delete?

File: tools/compliance/sbom.bzl [hunk 3]
```diff
@@ -27,6 +27,12 @@ def _sbom_impl(ctx):
     inputs = [inps[0]]
     args.add("--packages_used", inps[0].path)
     args.add("--out", ctx.outputs.out.path)
+    if ctx.attr._maven_install:
```
Comment (sdtwigg) [inline tools/compliance/sbom.bzl:30]: won't this always be true b/c it has a default set?

File: tools/compliance/write_sbom.py [hunk 4]
```diff
@@ -67,21 +70,40 @@ def create_sbom(package_info: dict) -> dict:  # pylint: disable=g-bare-generic
...
+    id = hashlib.md5()
+    id.update(pkg.encode('utf-8'))
+    spdxid = 'SPDXRef-GooglePackage-%s' % id.hexdigest()
+    pi = {
+        "name": pkg,
+        "downloadLocation": "NOASSERTION",
+        "SPDXID": spdxid,
+        # TODO(aiuto): Fill in the rest
+        # "supplier": "Organization: Google LLC",
+        # "licenseConcluded": "License-da09db95a268defe",
```
Comment (sdtwigg) [inline tools/compliance/write_sbom.py:82]: minor nit:
if the License line is fake, I think should replace with License-###########

mostly to make it immediately clear that these lines are example placeholders versus arbitrarily commented out.

File: tools/compliance/write_sbom.py [hunk 5]
```diff
@@ -67,21 +70,40 @@ def create_sbom(package_info: dict) -> dict:  # pylint: disable=g-bare-generic
...
+
+    have_maven = None
+    if pkg.startswith('@maven//:'):
+      have_maven = maven_packages.get(pkg[9:])
+    elif pkg.endswith("//file:file"):
+      # Bazel hacks jvm_external to add //file:file as a target, then we depend
+      # on that rather than the correct thing.
+      # Example: @org_apache_tomcat_tomcat_annotations_api_8_0_5//file:file
+      # Check for just the versioned root
+      have_maven = maven_packages.get(pkg[1:-11])
```
Comment (sdtwigg) [inline tools/compliance/write_sbom.py:94]: Python style is usually to do something like -len("//file:file") rather than have -11 as a magic number.

**Review by sdtwigg — CHANGES_REQUESTED** [review]
I left some initial cleanup comments on narrow issues I noticed while doing the first readthrough but I will also want to do another readthrough where I approach more holistically to look for wider issues.

File: tools/compliance/gather_packages.bzl (hunk 1 above)
Comment (aiuto) [reply tools/compliance/gather_packages.bzl:163]: Done

File: tools/compliance/sbom.bzl (hunk 2 above)
Comment (aiuto) [reply tools/compliance/sbom.bzl:50]: Done

File: tools/compliance/sbom.bzl (hunk 3 above)
Comment (aiuto) [reply tools/compliance/sbom.bzl:30]: Sigh. Yes, but I wanted the ability to be able to run this in test most without providing the lock file.

File: tools/compliance/write_sbom.py (hunk 4 above)
Comment (aiuto) [reply tools/compliance/write_sbom.py:82]: Sure. Done

File: tools/compliance/write_sbom.py (hunk 5 above)
Comment (aiuto) [reply tools/compliance/write_sbom.py:94]: I moved the string out to a constant.

**Review by sdtwigg — APPROVED** [review]

**Pull request closed without merging**